Continue processing work item in case of error

tedyu opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
In Store#DeleteCollection, upon error, the goroutine would return, resulting in fewer number of workers and potentially leaving some unprocessed items in toProcess channel.

This PR continues processing item in case of error.

A test has been added for the fix. Validation is used to produce error for one of the Pods to be deleted. There can be other reasons where deletion may fail.

**Which issue(s) this PR fixes**:
Fixes #

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

tedyu commented:
/test pull-kubernetes-node-e2e-containerd

tedyu commented:
/cc @deads2k @liggitt @lavalamp

tedyu commented:
/test pull-kubernetes-e2e-gce

tedyu commented:
/test pull-kubernetes-e2e-gce

tedyu commented:
/priority backlog

haosdent commented:
+1

tedyu commented:
This is a bug fix.
ping @deads2k @liggitt @lavalamp for review.

liggitt commented:
The current behavior of workers exiting on first error is working as intended... we can discuss whether that is still what we want or not. Errors in these code paths are not expected, so exiting early is not expected to be problematic. Has this actually been observed in real clusters in a way that this change would have helped?

liggitt commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1036,12 +1036,11 @@ func (e *Store) DeleteCollection(ctx context.Context, deleteValidation rest.Vali
 				accessor, err := meta.Accessor(items[index])
 				if err != nil {
 					errs <- err
```
by not exiting early, this means a single worker can enqueue multiple errors, and block on a full errs channel

tedyu commented:
Please refer to the test case: podA and podC are symmetrical. It is unexpected that podA is deleted while podC is not.
There was no indication of such behavior (dangling Pod) in the current error mechanism.

tedyu commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

tedyu commented:
/test pull-kubernetes-node-e2e-containerd

tedyu commented:
/test pull-kubernetes-e2e-kind

tedyu commented:
/test pull-kubernetes-e2e-gce

liggitt commented:
/assign

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1035,13 +1035,15 @@ func (e *Store) DeleteCollection(ctx context.Context, deleteValidation rest.Vali
 				if err != nil {
-					errs <- err
-					return
+					select {
```
I don't like this change because it means that some errors can end up unreported.  The buffer is only as big as the number of workers, not the amount of work.

This also reveals another problem with the method: not all errors are reported, not even numbers of errors.

On the practical side, this is a very unlikely error to get.  We care more about the actual deletion errors. below.

tedyu replied:
Adopted error reporting summary suggested below.

lavalamp replied:
You cannot use a select here, that drops errors on the floor.

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1035,13 +1035,15 @@ func (e *Store) DeleteCollection(ctx context.Context, deleteValidation rest.Vali
 				if _, _, err := e.Delete(ctx, accessor.GetName(), deleteValidation, options); err != nil && !apierrors.IsNotFound(err) {
 					klog.V(4).Infof("Delete %s in DeleteCollection failed: %v", accessor.GetName(), err)
 					errs <- err
-					return
```
I haven't read down all the way yet, but the only reason I can think that you didn't build a similar select structure here is that your unit test never checks this branch.

tedyu replied:
I run the new test which hits this line:
```
k8s.io/apiserver/pkg/registry/generic/registry.(*Store).DeleteCollection.func2 at store.go:1061
runtime.goexit at asm_amd64.s:1357
 - Async stack trace
k8s.io/apiserver/pkg/registry/generic/registry.(*Store).DeleteCollection at store.go:1028
```

deads2k commented:
I'm ok with the idea if the error reporting is correct first to handle multiple errors with some kind of summarization after the first X errors

tedyu commented:
@deads2k 
Thanks for taking a look.

I have modified error reporting to adopt your suggestion above.
If you prefer a different error reporting threshold, please let me know.

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1031,17 +1031,34 @@ func (e *Store) DeleteCollection(ctx context.Context, deleteValidation rest.Vali
 				if err != nil {
-					errs <- err
-					return
+					errorsSeen++
```
this is the wrong spot to try to aggregate errors. Errors are getting aggregated below after the wait group completes, that structure makes multi-error reporting per go-func risky for deadlock.  That's the area that needs fixing. Not this loop.

tedyu replied:
You're suggesting to do error aggregation around line 1067:
```
	select {
	case err := <-errs:
```
I would do this in next push where error count and last error is returned.

tedyu commented:
@deads2k 
Please take another look.

tedyu commented:
/test pull-kubernetes-e2e-gce

tedyu commented:
@deads2k 
I have modified TestStoreDeleteCollectionWithValidation to exercise case where two errors are encountered during deletion.
Added assertion for the error count.

tedyu commented:
/test pull-kubernetes-e2e-gce

tedyu commented:
/test pull-kubernetes-e2e-gce

yutedz committed:
Continue processing work item in case of error

tedyu commented:
/test pull-kubernetes-node-e2e-containerd

lavalamp commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1035,24 +1035,37 @@ func (e *Store) DeleteCollection(ctx context.Context, deleteValidation rest.Vali
-		return nil, err
-	default:
-		return listObj, nil
+	errorsSeen := 0
```
collecting the errors after all goroutines have finished requires sizing the errors channel assuming everything fails. I don't think that's a good idea.

lavalamp commented:
What is the purpose of this PR? if there's any error at all the operation will have to be retried, so it doesn't seem like this would reduce wall time under almost any imaginable circumstance.

tedyu commented:
@lavalamp 
Can you take a look at https://github.com/kubernetes/kubernetes/pull/86524#issuecomment-569118602 and give your opinion ?

lavalamp commented:
No matter which pod fails to be deleted, the caller is going to retry the delete collection call. It is not needed, desired, or even possible to list all the non-deleted objects in the error message (since objects can be added by others in a race). No user has complained about this to my knowledge.

tedyu commented:
From my first pass of the callers of DeleteCollection(), I don't see retry - please let me know if I miss something.

In the current code, even retry may not achieve what the caller intends (depending on the value of workersNumber) because the same object which didn't pass deletion in prior call may be attempted and fails again.

lavalamp commented:
The namespace deleter is just going to keep trying until it works.

On Wed, Jan 8, 2020 at 11:07 AM Ted Yu <notifications@github.com> wrote:

> From my first pass of the callers of DeleteCollection(), I don't see retry
> - please let me know if I miss something.
>
> In the current code, even retry may not achieve what the caller intends
> (depending on the value of workersNumber) because the same object which
> didn't pass deletion in prior call may be attempted and fails again.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/86524?email_source=notifications&email_token=AAE6BFVJ3QA2H4ONDZQHFRLQ4YP73A5CNFSM4J6ML7C2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEINUFFI#issuecomment-572211861>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE6BFXKWUUJIRTG36J62HDQ4YP73ANCNFSM4J6ML7CQ>
> .
>

tedyu commented:
We can revisit this when there is user issue surfacing.
